pywhy_graphs/algorithms/pag.py

```diff
@@ -1176,3 +1187,190 @@ def pag_to_mag(graph):
+    """Checks if the provided graph is a valid Partial ancestral graph (MAG).
+
+    A valid PAG as defined in :footcite:`Zhang2008` is a mixed edge graph that
+    has no directed or almost directed cycles and no inducing paths between
+    any two non-adjacent pair of nodes.
+
+    Parameters
+    ----------
+    G : Graph
+        The graph.
+
+    Returns
+    -------
+    is_valid : bool
+        A boolean indicating whether the provided graph is a valid PAG or not.
+
+    """
+
+    if L is None:
+        L = set()
+
+    if S is None:
+        S = set()
+
+    directed_sub_graph = G.sub_directed_graph()
+
+    all_nodes = set(G.nodes)
+
+    # check if there are more than one edges b/w two nodes
+    for node in all_nodes:
+        nb = set(G.neighbors(node))
+        for elem in nb:
+            edge_data = G.get_edge_data(node, elem)
+            if (edge_data["bidirected"] is not None) and (edge_data["directed"] is not None):
+                return False
+
+    # check if there are any directed cyclces
+    try:
+        nx.find_cycle(directed_sub_graph)  # raises a NetworkXNoCycle error
+        return False
+    except nx.NetworkXNoCycle:
+        pass
+
+    # check if there are any almost directed cycles
+    if has_adc(G):  # if there is an ADC, it's not a valid MAG
+        return False
+
+    # check if there are any inducing paths between non-adjacent nodes in the non-circle edge sub-graph
+
+    dedges = list(G.edges()["directed"])
+    # undedges = list(G.edges()["undirected"])
+    biedges = list(G.edges()["bidirected"])
+
+    temp_pag = PAG()
+
+    temp_pag.add_edges_from(dedges, temp_pag.directed_edge_name)
+
+    # can't remember why I only handle directed and bidirected edges
+
+    # temp_pag.add_edges_from(undedges, temp_pag.undirected_edge_name)
+
+    temp_pag.add_edges_from(biedges, temp_pag.bidirected_edge_name)
+
+    all_nodes = set(temp_pag.nodes)
+
+    for source in all_nodes:
+        nb = set(temp_pag.neighbors(source))
+        cur_set = all_nodes - nb
+        cur_set.remove(source)
+        for dest in cur_set:
+            out = inducing_path(temp_pag, source, dest, L, S)
+            if out[0] is True:
+                return False
+
+    return True
+
+
+def mag_to_pag(G: PAG):
+    """Converts the provided mag into a pag using the FCI algorithm.
+    The FCI algorithms, as defined in :footcite:`Zhang2008` is a provably
+    complete for learning all the tractable features of an MAG, thus
+    producing a PAG.
+
+    Parameters
+    ----------
+    G : MAG
+        The MAG.
+
+    Returns
+    -------
+    pag : PAG
+        The PAG constructed from the MAG.
+    """
+
+    data = dummy_sample(G)
+    oracle = Oracle(G)
+    # ci_estimator = GSquareCITest(data_type="discrete")
+    context = make_context().variables(data=data).build()
+    fci = FCI(ci_estimator=oracle)
+    fci.learn_graph(data, context)
+
+    return fci.graph_
+
+
+def equivalent_pag(G1: PAG, G2: PAG):
```

Comment:
This can probably be a private function and since it is only checking the equivalence of edges, you can probably rename to: `_check_pag_edges_are_equal`